File: Task 1/C++/Components/TransportManager.cpp
#include "TransportManager.h"

TransportManager::TransportManager() {}

void TransportManager::addTransport(Transport* newTransport) {
    if(newTransport == nullptr){return;}

    if(getTransport(newTransport->getCode()) == nullptr){ // check if created properly and not duplicate
        transportList.push_back(newTransport);
        printf("\nTransport Added\n");
        return;
    }
    printf("\nERROR: Code Already in Use\n");
}

void TransportManager::removeTransport(int code) {
    for(int i = 0; i < transportList.size(); i++){

        if(transportList[i]->getCode() == code){
            delete transportList[i];
            transportList.erase(transportList.begin()+i);
            printf("\nDeleted Successfully\n");
            return;
        }
    }
    printf("\nERROR: Code not found\n");
}

void TransportManager::updateTransport(int code, Transport* newTransport){
    if(newTransport == nullptr){return;}

    Transport* toUpdate = getTransport(code);
    if (toUpdate != nullptr) {

        // Check that Transport either has the same code as the one being updated or a different code
        if ((newTransport->getCode() == toUpdate->getCode()) || 
            (newTransport->getCode() != toUpdate->getCode() && getTransport(newTransport->getCode()) == nullptr)) {

            removeTransport(code);
            addTransport(newTransport);
            printf("\nTransport Updated\n");
            return;
        }
        printf("Transport was NOT Updated.\n");
    } else {
        printf("Transport not found in Transport list.\n");
    }
}


Transport* TransportManager::getTransport(int code){
    for(Transport* t : transportList){
        if(t->getCode() == code){
            return t;
        };
    }
    return nullptr;
}


void TransportManager::displayTransports(){
    printf("\n===========Transport List===========\n");
    for (Transport* transport : transportList) {
        printf("\n%s",transport->to_string().c_str());
    }
}
void TransportManager::displayTransports(int code){
    Transport* transport = getTransport(code);
    if(transport != nullptr){
        printf("\n%s",transport->to_string().c_str());
        return;
    }
    printf("\nERROR: Code not found\n");
}

void TransportManager::clearTransport(){
    for (Transport* transport : transportList) {
        delete transport;
    }
    transportList.clear();
}


TransportManager::~TransportManager() {
    clearTransport();
}

File: Task 2/Java/logistics/src/main/java/com/logistics/Components/TransportManager.java
package com.logistics.Components;
import java.util.ArrayList;
import java.util.List;

import com.logistics.Components.Transport.Transport;

public class TransportManager {
    private List<Transport> transportList;

    // Default constructor
    public TransportManager() {
        transportList = new ArrayList<>();
    }

    // Adds a new transport to the transport list
    public void addTransport(Transport newTransport) {
        if (newTransport == null) {
            return;
        }

        if (getTransport(newTransport.getCode()) == null) {
            transportList.add(newTransport);
            System.out.println("\nTransport Added\n");
            return;
        }

        System.out.println("\nERROR: Code Already in Use\n");
    }

    // Removes a transport from the transport list based on its code
    public void removeTransport(int code) {
        for (int i = 0; i < transportList.size(); i++) {
            if (transportList.get(i).getCode() == code) {
                transportList.remove(i);
                System.out.println("\nDeleted Successfully\n");
                return;
            }
        }
        System.out.println("\nERROR: Code not found\n");
    }

    // Updates a transport in the transport list based on its code
    public void updateTransport(int code, Transport newTransport) {
        if (newTransport == null) {
            return;
        }

        Transport toUpdate = getTransport(code);
        if (toUpdate != null) {
            // Check that the transport either has the same code as the one being updated or a different code
            if (newTransport.getCode() == toUpdate.getCode() ||
                    (newTransport.getCode() != toUpdate.getCode() && getTransport(newTransport.getCode()) == null)) {
                removeTransport(code);
                addTransport(newTransport);
                System.out.println("\nTransport Updated\n");
                return;
            }
            System.out.println("Transport was NOT Updated.\n");
        } else {
            System.out.println("Transport not found in Transport list.\n");
        }
    }

    // Retrieves a transport from the transport list based on its code
    public Transport getTransport(int code) {
        for (Transport t : transportList) {
            if (t.getCode() == code) {
                return t;
            }
        }
        return null;
    }

    // Displays all transports in the transport list
    public void displayTransports() {
        System.out.println("\n===========Transport List===========\n");
        for (Transport transport : transportList) {
            System.out.println("\n" + transport.to_string());
        }
    }

    // Displays a specific transport from the transport list based on its code
    public void displayTransports(int code) {
        Transport transport = getTransport(code);
        if (transport != null) {
            System.out.println("\n" + transport.to_string());
            return;
        }
        System.out.println("\nERROR: Code not found\n");
    }

    // Clears the transport list and manages memory
    public void clearTransport() {
        transportList.clear();
    }
}
